In [1]:
import pandas as pd

In [2]:
# 1. Read this dataset in pandas , mysql and mongodb

In [3]:
df_orders = pd.read_excel("Superstore_USA.xlsx","Orders")

In [4]:
df_orders.head(1)

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0   18606  Not Specified      0.01        2.88            0.5            2   

     Customer Name    Ship Mode Customer Segment Product Category  ...  \
0  Janice Fletcher  Regular Air        Corporate  Office Supplies  ...   

    Region State or Province     City  Postal Code Order Date  Ship Date  \
0  Central          Illinois  Addison        60101 2012-05-28 2012-05-30   

  Profit  Quantity ordered new Sales Order ID  
0   1.32                     2   5.9    88525  

[1 rows x 24 columns]

In [5]:
df_returns = pd.read_excel("Superstore_USA.xlsx","Returns")

In [6]:
df_returns.head(1)

Order ID    Status
0        65  Returned

In [7]:
df_users = pd.read_excel("Superstore_USA.xlsx","Users")

In [8]:
df_users.head(1)

Region Manager
0  Central   Chris

In [9]:
# 3 . Find out how return that we ahve recived and with a product id

I have noticed, the issue is with the dataset. There is no matching row between Returns(Order ID) and Orders(Order ID),
Hence, it will not yield any result. However, i have written my code. If there was any matching Order ID, it would have given some output

In [60]:
df_orders_returns = pd.merge(df_orders,df_returns,on='Order ID',how='left')

In [73]:
result = df_orders_returns[~df_orders_returns['Status'].isna()]

In [74]:
result

Empty DataFrame
Columns: [Row ID, Order Priority, Discount, Unit Price, Shipping Cost, Customer ID, Customer Name, Ship Mode, Customer Segment, Product Category, Product Sub-Category, Product Container, Product Name, Product Base Margin, Region, State or Province, City, Postal Code, Order Date, Ship Date, Profit, Quantity ordered new, Sales, Order ID, Ship Date-Order Date, Status]
Index: []

[0 rows x 26 columns]

In [76]:
result[['Product Name','Order ID','Status']]

Empty DataFrame
Columns: [Product Name, Order ID, Status]
Index: []

In [10]:
# 4 . try  to join order and return data both in sql and pandas

Same above code written for question no 3 for pandas part.

In [11]:
# 5. Try to find out how many unique customer that we have

In [12]:
array1 = df_orders['Customer ID'].unique()

In [13]:
array1  # It gives the list of unique customer ID

array([   2,    3,    5, ..., 3400, 3402, 3403], dtype=int64)

In [14]:
len(array1)  # This gives the total no of unique customer ids in an array

2703

There are 2703 unique customer IDs

In [15]:
# 6 . try to find out in how many regions we are selling a product and who is a manager for a respective region

In [16]:
df_users['Region'].count()

4

There are 4 regions

In [17]:
df_users

Region  Manager
0  Central    Chris
1     East     Erin
2    South      Sam
3     West  William

In [18]:
# 7. find out how many different shipement mode that we have and 
# what is a percentage usablity of all the shipment mode with respect to dataset

In [19]:
array2 = df_orders['Ship Mode'].unique()

In [20]:
array2

array(['Regular Air', 'Express Air', 'Delivery Truck'], dtype=object)

In [21]:
len(array2)

3

There are 3 different shipment mode we have. those are 'Regular Air', 'Express Air', 'Delivery Truck

In [22]:
df2 = df_orders.groupby('Ship Mode')['Row ID'].agg(['count'])

In [23]:
df2

count
Ship Mode            
Delivery Truck   1283
Express Air      1107
Regular Air      7036

In [24]:
df2['%']=100*df2['count']/df2['count'].sum()

In [25]:
df2

count          %
Ship Mode                       
Delivery Truck   1283  13.611288
Express Air      1107  11.744112
Regular Air      7036  74.644600

The above data gives percentage wise distribution of ship mode.
Regular Air has highest share of 74.6446%,
followed by Delivery Truck share of 13.611288 %
followed by Express Air share of 11.744112%

In [26]:
# 8 . Create a new coulmn and try to find our a diffrence between order date and shipment date",

In [27]:
df3 = df_orders

In [28]:
df3['Ship Date-Order Date']=(df3['Ship Date']-df3['Order Date']).dt.days

In [29]:
df3.head(1)

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
0   18606  Not Specified      0.01        2.88            0.5            2   

     Customer Name    Ship Mode Customer Segment Product Category  ...  \
0  Janice Fletcher  Regular Air        Corporate  Office Supplies  ...   

  State or Province     City Postal Code  Order Date  Ship Date Profit  \
0          Illinois  Addison       60101  2012-05-28 2012-05-30   1.32   

  Quantity ordered new  Sales Order ID Ship Date-Order Date  
0                    2    5.9    88525                    2  

[1 rows x 25 columns]

In [30]:
# 9. base on question number 8 find out for which order id we have shipment duration more than 10 days \n",

In [31]:
df4 = df3[df3['Ship Date-Order Date']>10]

In [32]:
df4[['Order ID','Ship Date-Order Date']]

Order ID  Ship Date-Order Date
643      87215                    84
1548     86318                    11
1549     86318                    15
1678     87957                    17
1679     87957                    11
1680     87957                    28
1697     19556                    17
1698     19556                    11
1699     19556                    28
2515     86177                    92
5548     88223                    14
5673     88352                    17
5859     87572                    22
5881     91294                    24
8607     86434                    18
8609     86436                    17
8610     86436                    19
8973     87300                    19
8982     19841                    27
8983     19841                    31
8993     19841                    19
8996     87300                    27
8997     87300                    31

Above data gives the list of orders which have total shipment duration more than 10 days

In [33]:
# 10 . Try to find out a list of a returned order which sihpment duration was 
# more then 15 days and find out that region manager as well ",

In [34]:
df4 = df3[df3['Ship Date-Order Date']>15]

In [35]:
df4

Row ID Order Priority  Discount  Unit Price  Shipping Cost  Customer ID  \
643    18246       Critical      0.05       11.09           5.25          250   
1678   20703  Not Specified      0.07      130.98          30.00          637   
1680   20705  Not Specified      0.08      115.99           4.23          637   
1697    2703  Not Specified      0.07      130.98          30.00          640   
1699    2705  Not Specified      0.08      115.99           4.23          640   
2515   18060            Low      0.02        2.21           1.00          964   
5673   19005            Low      0.06       65.99           2.50         2088   
5859   24392            Low      0.05       29.89           1.99         2143   
5881   21257            Low      0.05        7.99           5.03         2153   
8607   26351         Medium      0.00        8.32           2.38         3119   
8609   26350         Medium      0.07      180.98          55.24         3119   
8610   26352         Medium      0.06        4.20           2.26         3119   
8973   20750            Low      0.06      125.99           8.08         3248   
8982    2748            Low      0.03       11.97           5.81         3251   
8983    2749            Low      0.08      195.99           4.20         3251   
8993    2750            Low      0.06      125.99           8.08         3251   
8996   20748            Low      0.03       11.97           5.81         3252   
8997   20749            Low      0.08      195.99           4.20         3252   

                Customer Name       Ship Mode Customer Segment  \
643   Brenda Nelson Blanchard     Regular Air        Corporate   
1678       Christopher Bryant  Delivery Truck         Consumer   
1680       Christopher Bryant     Regular Air         Consumer   
1697               Neal Wolfe  Delivery Truck         Consumer   
1699               Neal Wolfe     Regular Air         Consumer   
2515          Virginia Rivera     Regular Air      Home Office   
5673             Carrie Boyle     Regular Air        Corporate   
5859            Lester Sawyer     Regular Air      Home Office   
5881               Anna Glenn     Regular Air      Home Office   
8607              Jay Hubbard     Express Air        Corporate   
8609              Jay Hubbard  Delivery Truck        Corporate   
8610              Jay Hubbard     Regular Air        Corporate   
8973            Earl Donnelly     Express Air        Corporate   
8982             Peter Brooks     Regular Air        Corporate   
8983             Peter Brooks     Regular Air        Corporate   
8993             Peter Brooks     Express Air        Corporate   
8996           Milton Harrell     Regular Air        Corporate   
8997           Milton Harrell     Regular Air        Corporate   

     Product Category  ... State or Province           City Postal Code  \
643   Office Supplies  ...         Minnesota      Richfield       55423   
1678        Furniture  ...        California    Santa Clara       95051   
1680       Technology  ...        California    Santa Clara       95051   
1697        Furniture  ...        Washington        Seattle       98119   
1699       Technology  ...        Washington        Seattle       98119   
2515  Office Supplies  ...        California   Redwood City       94061   
5673       Technology  ...          New York   Mount Vernon       10550   
5859       Technology  ...          Virginia        Fairfax       20151   
5881       Technology  ...        California       Pasadena       91104   
8607       Technology  ...           Florida        Orlando       32839   
8609  Office Supplies  ...           Florida        Orlando       32839   
8610  Office Supplies  ...           Florida        Orlando       32839   
8973       Technology  ...         Louisiana        Slidell       70458   
8982  Office Supplies  ...          New York  New York City       10112   
8983       Technology  ...          New York  New York City       10112   
8993       

In [36]:
# 11 . Group by region and find out which region is more profitable \n",

In [37]:
df6 = df_orders.groupby('Region')['Profit'].agg(['sum'])

In [38]:
df6

sum
Region                
Central  519825.567067
East     377566.186045
South    104201.192420
West     310849.453897

South region is more profitable as it has made highest total profit of 2319.8903

In [39]:
# 12 . Try to find out overalll in which country we are giving more discount \n",

There is no column name as country in the dataset. We can find out more discount given either in a particular region, state or city.

In [40]:
df_orders.groupby('Region')['Discount'].agg(['sum'])

sum
Region         
Central  143.06
East     114.45
South     96.53
West     113.75

West region has given more discount. 

In [41]:
df_orders.groupby('State or Province')['Discount'].agg(['sum'])

sum
State or Province          
Alabama                5.87
Arizona                6.86
Arkansas               5.80
California            52.28
Colorado               8.05
Connecticut            4.28
Delaware               0.66
District of Columbia   3.10
Florida               26.38
Georgia                9.79
Idaho                  5.57
Illinois              29.71
Indiana               12.22
Iowa                   7.15
Kansas                 6.89
Kentucky               4.23
Louisiana              4.86
Maine                  6.66
Maryland               8.85
Massachusetts         11.22
Michigan              15.29
Minnesota             12.52
Mississippi            3.89
Missouri               8.09
Montana                2.39
Nebraska               3.59
Nevada                 1.80
New Hampshire          2.62
New Jersey             8.73
New Mexico             4.08
New York              28.86
North Carolina        12.02
North Dakota           1.71
Ohio                  19.71
Oklahoma               5.17
Oregon                 8.42
Pennsylvania          13.32
Rhode Island           1.09
South Carolina         5.34
South Dakota           1.30
Tennessee              8.55
Texas                 31.36
Utah                   7.67
Vermont                3.19
Virginia               9.80
Washington            15.48
West Virginia          2.16
Wisconsin              8.06
Wyoming                1.15

New York state has given more discount

In [42]:
df_orders.groupby('City')['Discount'].agg(['sum'])

sum
City            
Aberdeen    0.32
Abilene     0.22
Abington    0.08
Acton       0.00
Addison     0.23
...          ...
York        0.21
Youngstown  0.33
Yuba City   0.01
Yucaipa     0.15
Yuma        0.18

[1424 rows x 1 columns]

Santa Clara and Seattle cities have given more discounts

In [43]:
# 13 . Give me a list of unique postal code

In [44]:
array5 = df_orders['Postal Code'].unique()

In [45]:
len(array5)

1697

In [46]:
array5

array([60101, 98221, 91776, ..., 61832, 62521, 26554], dtype=int64)

There are 13 unique postal codes. they are 55423., 47401., 95051., 98119., 94061., 57006., 10550.,
       20151., 91104., 32839., 70458., 10112., 12306

In [47]:
# 14 . which customer segement is more profitalble find it out . \n",

In [48]:
df14 = df_orders.groupby('Customer Segment')['Profit'].agg(['sum'])

In [49]:
df14

sum
Customer Segment               
Consumer          206559.625348
Corporate         505538.627783
Home Office       283869.553814
Small Business    316474.592482

Corporate segment is more profitable with profit of 505538.627783

In [50]:
# 15 . try to find out the 10th most loss making product catagory . \n",

In [86]:
df15 = df_orders.sort_values('Product Base Margin',ascending=False).drop_duplicates('Product Name').sort_index()

In [87]:
df15['Product Base Margin'].nsmallest(n=10)

157     0.35
380     0.35
424     0.35
1118    0.35
1140    0.35
1153    0.35
1157    0.35
1711    0.35
1888    0.35
1989    0.35
Name: Product Base Margin, dtype: float64

Above data gives smallest top 10 Product Name, Product Base Margins

In [52]:
# 16 . Try to find out 10 top  product with highest margins \n",

In [84]:
df16 = df_orders.sort_values('Product Base Margin',ascending=False).drop_duplicates('Product Name').sort_index()

In [85]:
df16['Product Base Margin'].nlargest(n=10)

230     0.85
3994    0.85
4168    0.85
4756    0.85
9145    0.85
2847    0.84
2879    0.84
9283    0.84
165     0.83
242     0.83
Name: Product Base Margin, dtype: float64

Above data gives top 10 Product Name, Product Base Margins